File: Python/string-1/make_abba.py
'''
CodingBat Python Activity "make_abba" from string-1.
codingbat.com
'''

failures = 0

def make_abba(a, b):
    '''
Given two strings, a and b, return the result of putting them together in the order abba,
e.g. "Hi" and "Bye" returns "HiByeByeHi".


make_abba('Hi', 'Bye') → 'HiByeByeHi'
make_abba('Yo', 'Alice') → 'YoAliceAliceYo'
make_abba('What', 'Up') → 'WhatUpUpWhat'
'''
    pass #Replace pass with your code to implement this function





def catchFailures(testCase, result):
    global failures
    
    try:
        assert eval(testCase) == result, "Test Case Failed: "+testCase+ " → "+str(result)
    except Exception as e:
        failures += 1
        print(e)
            


def test():
    global failures
    
    catchFailures("make_abba('Hi', 'Bye')", 'HiByeByeHi')
    catchFailures("make_abba('Yo', 'Alice')", 'YoAliceAliceYo')
    catchFailures("make_abba('What', 'Up')", 'WhatUpUpWhat')
    catchFailures("make_abba('aaa', 'bbb')", 'aaabbbbbbaaa')
    catchFailures("make_abba('x', 'y')", 'xyyx')
    catchFailures("make_abba('x', '')", 'xx')
    catchFailures("make_abba('', 'y')", 'yy')
    catchFailures("make_abba('Bo', 'Ya')", 'BoYaYaBo')
    catchFailures("make_abba('Ya', 'Ya')", 'YaYaYaYa')

        
    if failures > 0:
        print("\r\n "+ str(failures) + " failed test case(s)   :(")
    else:
        print("Good job -- problem solved!")


test()

File: Java/String1/MakeAbba.java
/**
 * CodingBat Java Activity "makeAbba" from String1.
 * codingbat.com
 */
public class MakeAbba
{
	private static int failures = 0;
	private static final String METHOD_NAME = "makeAbba";

	/**
	 * INSTRUCTIONS:
	 * Given two strings, a and b, return the result of putting them together 
	 * in the order abba, e.g. "Hi" and "Bye" returns "HiByeByeHi".
	 * 
	 * makeAbba("Hi", "Bye") -> 19
	 * makeAbba("Yo", "Alice") -> "YoAliceAliceYo"
	 * makeAbba("What", "Up") -> "WhatUpUpWhat"
	 * 
	 */
	public static String makeAbba(String a, String b)
	{
		// ADD YOUR CODE HERE
	}

	public static void main(String[] args)
	{
		testCase("Hi", "Bye", "HiByeByeHi");
		testCase("Yo", "Alice", "YoAliceAliceYo");
		testCase("What", "Up", "WhatUpUpWhat");
		testCase("aaa", "bbb", "aaabbbbbbaaa");
		testCase("x", "y", "xyyx");
		testCase("x", "", "xx");
		testCase("", "y", "yy");
		testCase("Bo", "Ya", "BoYaYaBo");
		testCase("Ya", "Ya", "YaYaYaYa");
		
		System.out.println();
		System.out.println(failures + " failed test cases.");
	}

	private static void testCase(String a, String b, String result)
	{
		String r;
		try
		{
			r = ""+makeAbba(a, b);
			if (r.equals(""+result))
			{
				System.out.print("PASS: ");
			}
			else
			{
				System.out.print("FAIL: ");
				failures ++;
			}
		}
		catch (Exception e)
		{
			r = e.getMessage();
			System.out.print("FAIL: ");
			failures ++;
		}
		System.out.println(METHOD_NAME+"(\""+a+"\", \""+b+"\") -> \""+r + "\", EXPECTED: \""+result+"\"");
	}

}
